### PR TITLE
Tests: object api multipart tests and bug fixes.

### PR DESCRIPTION
**Pull request by hackintoshrao**

### CODE REVIEW

File: fs.go [hunk 1]
```diff
@@ -17,6 +17,8 @@
 package main
 
 import (
+	"github.com/minio/minio/pkg/disk"
+	"github.com/minio/minio/pkg/safe"
```
Comment (harshavardhana) [inline fs.go]: Looks like 'gofmt' is missing on this part of source.

File: fs.go [hunk 2]
```diff
@@ -617,10 +631,13 @@ func (s fsStorage) CreateFile(volume, path string) (writeCloser io.WriteCloser,
 func (s fsStorage) StatFile(volume, path string) (file FileInfo, err error) {
 	volumeDir, err := s.getVolumeDir(volume)
 	if err != nil {
+<<<<<<< HEAD
```
Comment (harshavardhana) [inline fs.go]: Conflict commits pushed 👎

File: fs.go [hunk 3]
```diff
@@ -74,6 +76,18 @@ func isDirExist(dirname string) (bool, error) {
 	return fi.IsDir(), nil
 }
 
+func (s fsStorage) isBucketExists(bucketName string) (isExist bool, err error) {
```
Comment (harshavardhana) [inline fs.go]: storage shouldn't have a concept of buckets, it should be named  - `isVolumeExists(volume string)`

File: object-api-multipart.go [hunk 4]
```diff
@@ -347,11 +380,22 @@ func (o objectAPI) PutObjectPart(bucket, object, uploadID string, partID int, si
...
+		// Allocating a byte to read from the reader.
+		// This validate whether more data exists even after 'size' number of bytes.
+		singleByte := make([]byte, 1)
+		singleByteBuffer := bytes.NewBuffer(singleByte)
+		// Reader shouldn't have more data what mentioned in size arguement.
+		// reading one more byte from the reader to validate it.
+		// expected to fail, success validates existance of more data in the reader.
+		if _, e = io.CopyN(singleByteBuffer, data, 1); e == nil {
+			safeCloseAndRemove(fileWriter)
+			return "", probe.NewError(UnExpectedDataSize{Message: fmt.Sprintf("Contains more data then specified size of %d bytes.", size)})
```
Comment (harshavardhana) [inline object-api-multipart.go]: The message should go into the Error() string in the definition itself. Not here.

File: object-api-multipart.go [hunk 5]
```diff
@@ -347,11 +380,22 @@ func (o objectAPI) PutObjectPart(bucket, object, uploadID string, partID int, si
...
 			safeCloseAndRemove(fileWriter)
-			if e == io.ErrUnexpectedEOF {
+			if e == io.ErrUnexpectedEOF || e == io.ErrShortWrite {
 				return "", probe.NewError(IncompleteBody{})
 			}
 			return "", probe.NewError(e)
 		}
+		// Allocating a byte to read from the reader.
+		// This validate whether more data exists even after 'size' number of bytes.
+		singleByte := make([]byte, 1)
```
Comment (harshavardhana) [inline object-api-multipart.go]: To read a single byte you don't need to do this. You can just do the following, avoids allocating a byte 

```
if _, e = io.CopyN(ioutil.Discard, data, 1); e == nil {
```

Comment (hackintoshrao) [inline object-api-multipart.go]: Done.

File: object-api-multipart.go (hunk 4 above)
Comment (hackintoshrao) [inline object-api-multipart.go]: Done.

File: fs.go (hunk 2 above)
Comment (hackintoshrao) [inline fs.go]: Sorry about that. Messed it up.

**Comment by hackintoshrao** [comment]
Fixes #1370 and #1371 .

**Pull request merged**